Microchip I2C driver support for MEC family

albertofloyd opened the pull request:
This patch introduces the  I2C driver for MEC family.
Driver implements byte-to-byte data transfer using bitbang register from Microchip.
No interrupts required for bitbang, register polling is used with timeout according to I2C spec. 
DMA functionality could be implemented in the future.

albertofloyd committed:
ext: hal: mchp: Fix compilation for MEC1501 SMB macros

Adjust space definition for I2C register base addresses

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd committed:
dts: mchp: Add Microchip I2C SoC device tree

Expose MEC1501 5 I2C/SMB  controllers
Add Microchip specific I2C device tree properties

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd committed:
boards : arm : Add I2C support for MEC15xx EVB

Expose 3 I2C instances with default SoC-Phy mappings for use with EVB

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd committed:
CODEOWNERS: Add code ownership for Microchip XEC drivers

Set ownership for Microchip XEC drivers

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

andrewboie commented on drivers/i2c/i2c_mchp_xec.c:
```diff
@@ -0,0 +1,431 @@
+	return 0;
+}
+
+static int _wait_completion(u32_t ba)
```
identifiers should not be prefixed with _ that is reserved by the C standard

albertofloyd replied:
Done

andrewboie commented on drivers/i2c/i2c_mchp_xec.c:
```diff
@@ -0,0 +1,431 @@
+	},
+};
+
+static int counter;
```
I don't think this will work well with multiple driver instances doing things concurrently...shouldn't this be a member of device->driver_data instead?

albertofloyd replied:
Agree. function now receives a pointer like other drivers handling multiple instances

andrewboie commented on drivers/i2c/i2c_mchp_xec.c:
```diff
@@ -0,0 +1,431 @@
+	return 0;
+}
+
+static bool _check_lines(u32_t ba)
```
remove leading _

andrewboie commented on drivers/i2c/i2c_mchp_xec.c:
```diff
@@ -0,0 +1,431 @@
+{
+	counter = counter + 1;
+	if (counter > TIMEOUT) {
+		printk("Timeout!\n");
```
I think the printk should be removed, or use the logging mechanism so the user has some clue where the "Timeout!" comes from

franciscomunoz commented on drivers/i2c/i2c_mchp_xec.c:
```diff
@@ -0,0 +1,431 @@
+};
+DEVICE_AND_API_INIT(i2c_xec_0, DT_MICROCHIP_XEC_I2C_0_LABEL, &i2c_xec_init,
+			&i2c_xec_data_0, &i2c_xec_config_0,
+			POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,
```
Not sure about this one, but other i2C drivers are using the following priority 
CONFIG_KERNEL_INIT_PRIORITY_DEVICE. Can you comment @andrewboie ?

albertofloyd replied:
Seems to be a KConfig I2C-specific requirement, see other I2C drivers

albertofloyd replied:
@andrewboie could you comment on this priority, most of other request changes are done.

andrewboie replied:
I'm not sure either. This doesn't seem to be consistent across i2c drivers. I'm OK with leaving as-is.

franciscomunoz commented on dts/bindings/i2c/microchip,xec-i2c.yaml:
```diff
@@ -0,0 +1,60 @@
+version: 0.1
+
+description: >
+    This binding gives a base representation of  I2C/SMB controller for Microchip
```
for Microchip XEC

franciscomunoz commented on dts/bindings/i2c/microchip,xec-i2c.yaml:
```diff
@@ -0,0 +1,60 @@
+      generation: define
+      category: required
+
+    agg_irq:
```
Don't introduce unused features such as
agg_irq:
agg_irq_bit:
dir_irq:
is_aggr:

albertofloyd commented on drivers/i2c/i2c_mchp_xec.c:
```diff
@@ -0,0 +1,431 @@
+static const struct i2c_xec_config i2c_xec_config_0 = {
+	.base_addr = DT_MICROCHIP_XEC_I2C_0_BASE_ADDRESS,
+	.port_sel = DT_MICROCHIP_XEC_I2C_0_PORT_SEL,
+	.irq = DT_MICROCHIP_XEC_I2C_0_DIR_IRQ,
```
Removed

franciscomunoz commented on CODEOWNERS:
```diff
@@ -142,6 +142,7 @@
 /drivers/i2c/i2c_ll_stm32*                @ldts @ydamigos
 /drivers/i2c/i2c_rv32m1_lpi2c*            @henrikbrixandersen
 /drivers/i2c/*sam0*                       @Sizurka
+/drivers/i2c/*xec*                        @albertofloyd @scottwcpg
```
Change to  
/drivers/*/*chhp*                           @albertofloyd @scottwcpg @franciscomunoz 
we don't have create a CODEOWNERS commit for each new driver in the future

andrewboie replied:
agreed

scottwcpg commented:
1. HW I2C filter is not enabled in the Configuration register.  I2C spec. requires filtering.  Unless the board implements an external filter then Config.FEN should be set to 1.  A board level build config would work.
2. AAS and DSA slave address match and interrupt generation are enabled. This could also be a build config item if slave mode support is added in the future.
3. The MEC I2C controller spec. states bit-bang control should not be enabled if the controller's HW timeouts are enabled.  I suspect keeping bit-bang enabled only to read the lines should not be a problem.  Changing the lines with bit-bang while the HW timeouts are enabled could trigger a timeout.

albertofloyd committed:
soc: arm: mchp: Add I2C to device tree for MEC1501

Add additional support to device tree for I2C

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd committed:
drivers: i2c: mchp: Add I2C driver for MEC family

Add i2c_mchp_xec byte-to-byte driver to support I2C instances
for MEC family

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

albertofloyd commented:
> 1. HW I2C filter is not enabled in the Configuration register.  I2C spec. requires filtering.  Unless the board implements an external filter then Config.FEN should be set to 1.  A board level build config would work.
I2C filter enabled in the driver.
> 2. AAS and DSA slave address match and interrupt generation are enabled. This could also be a build config item if slave mode support is added in the future.
Removed I2C slave initialization.
> 3. The MEC I2C controller spec. states bit-bang control should not be enabled if the controller's HW timeouts are enabled.  I suspect keeping bit-bang enabled only to read the lines should not be a problem.  Changing the lines with bit-bang while the HW timeouts are enabled could trigger a timeout.
Added comment indicating HW timeouts are not be used in the driver.

franciscomunoz commented on drivers/i2c/Kconfig.xec:
```diff
@@ -0,0 +1,34 @@
+	depends on SOC_FAMILY_MEC
+	help
+	  This tells the driver to configure the I2C device at boot, depending
+	  on the additional configuration options below.
```
@albertofloyd This is a much better change. I also need to update my PR because it is also wrong.

# Kconfig.xec - Microchip XEC I2C configuration options
#
# Copyright (c) 2019 Intel Corporation
#
# SPDX-License-Identifier: Apache-2.0
#

menuconfig I2C_XEC
	bool "XEC Microchip I2C driver"
	depends on SOC_FAMILY_MEC
	select HAS_DTS_I2C
	help
	Enable the Microchip XEC I2C driver.

if I2C_XEC

config I2C_XEC_0
	bool "Enable I2C XEC device 0"
	help
	  This tells the driver to configure the I2C device at boot, depending
	  on the additional configuration options below.

config I2C_XEC_1
	bool "Enable I2C XEC device 1"
	help
	  This tells the driver to configure the I2C device at boot, depending
	  on the additional configuration options below.

config I2C_XEC_2
	bool "Enable I2C XEC device 2"
	help
	  This tells the driver to configure the I2C device at boot, depending
	  on the additional configuration options below.
	  
endif # I2C_XEC

franciscomunoz reviewed (commented):
The change proposed is aligned with the Kconfig in all the drivers.
